### PR TITLE
test: Add android detox coverage with new RN upgrade

### PR DESCRIPTION
**Pull request by cortisiko**
**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

**Description**

This PR enables Detox testing with Android on the react native upgrade PR. Initially, our test runs were limited to iOS devices, where we relied primarily on TestIDs to identify elements. However, when extending our test coverage to Android, we had to incorporate different detox [matchers](https://wix.github.io/Detox/docs/api/matchers/). To address these minor variations in platforms, we employ a straightforward IF-ELSE statement. We first verify whether the tests are executed on an iOS platform. If so, we utilize the `byId` matcher approach; otherwise, we resort to the `byLabel` matcher.

An example of this would be tapping on the confirm button on the transaction confirmation modal:

```
 if (device.getPlatform() === 'ios') {
      await TestHelpers.waitAndTap(CONFIRM_TRANSACTION_BUTTON_ID);
    } else {
      await TestHelpers.waitAndTapByLabel(CONFIRM_TRANSACTION_BUTTON_ID);
    }
```

In the above code block, we are verifying whether the current test is running on an iOS device. If it is an iOS device, we perform a tap action on the confirm button using the testID. However, if it is not an iOS device, we tap the confirm button byLabel.

### NOTE

- We have identified an ongoing visibility issue in React Native, which becomes more pronounced when running tests on devices with smaller screens. This issue is currently an open [ticket](https://github.com/facebook/react-native/issues/23870) in the react-native repo. Tests failed with a visibility error similar to `Test Failed: 15.0sec timeout expired without matching of given matcher: (view has effective visibility <VISIBLE> and view.getGlobalVisibleRect() covers at least <75> percent of the view's area)`
 
     - The recommendation here is to use pixel 5 or 6 devices when running tests. 
  
- There is currently an issue when running detox on android 12 and 13 (API 32/33) which prevents the tests from running. The issue is, the `tap()` action is treated like a `tapAndHold()` action. See the open issue in wix/detox [here](https://github.com/wix/Detox/issues/3762)
     
     -  The recommendation here is to use any android OS version lower than 12. API 31 or lower. Once detox addresses the issue with API 32/33 we will make the necessary adjustments on our end.  

- There are intermitted failures on bitrise due to network latency during the yarn setup step. See #6778 
       

   - If you do encounter this please rerun the build. 

- Detox has an intermittent crash on iOS which looks like:

```
Signal 11 was raised
(
	0   Detox                               0x0000000105e220f4 +[NSThread(DetoxUtils) dtx_demangledCallStackSymbols] + 36
	1   Detox                               0x0000000105e24d04 __DTXHandleCrash + 440
	2   Detox                               0x0000000105e25334 __DTXHandleSignal + 72
	3   libsystem_platform.dylib            0x00000001bca38560 _sigtramp + 52
	4   JavaScriptCore                      0x0000000108f6c824 WTF::RunLoop::TimerBase::start(WTF::Seconds, bool)::$_1::__invoke(__CFRunLoopTimer*, void*) + 52
	5   DetoxSync                           0x000000013fb5c5fc -[_DTXTimerTrampoline fire:] + 200
	6   DetoxSync                           0x000000013fb432fc _DTXCFTimerTrampoline + 68
	7   CoreFoundation                      0x000000010d198afc __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 28
	8   CoreFoundation                      0x000000010d19870c __CFRunLoopDoTimer + 1000
	9   CoreFoundation                      0x000000010d197c00 __CFRunLoopDoTimers + 324
	10  CoreFoundation                      0x000000010d192018 __CFRunLoopRun + 1868
	11  CoreFoundation                      0x000000010d1913bc CFRunLoopRunSpecific + 572
	12  MetaMask                            0x000000010269ec68 +[RCTCxxBridge runRunLoop] + 736
	13  DetoxSync                           0x000000013fb4b568 swz_runRunLoopThread + 276
	14  Foundation                          0x0000000107c13ca0 __NSThread__start__ + 848
	15  libsystem_pthread.dylib             0x00000001bca444e4 _pthread_start + 116
	16  libsystem_pthread.dylib             0x00000001bca3f6cc thread_start + 8

```
If you do encounter this error, try kicking off another test run. A short-term solution until this is addressed is to have test retries. 

   - There is currently an open issue in detox relating to this   https://github.com/wix/Detox/issues/3207#issuecomment-1618326655

### To trigger Android tests

Run this command:
`yarn test:e2e:android:debug --testNamePattern="Smoke"  `

**Screenshots/Recordings**

I was able to get tests running and passing on my local machine, see comment here: https://github.com/MetaMask/metamask-mobile/pull/6384#issuecomment-1589539720 

As well as bitrise:

https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/8ac3dc28-a4e7-4d17-be6a-7e4ccd244eba
https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/1a34585a-5f4b-48cc-8730-3282f2b89f93
https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/70280740-3d02-45f4-9e6e-b836738e176b
https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/d4d4cf3a-f6ed-40bf-9f1a-33e8b24a07de

**Issue**

Progresses #5913 

**Checklist**

* [ ] There is a related GitHub issue
* [ ] Tests are included if applicable
* [ ] Any added code is fully documented

### CODE REVIEW

**Comment by seaona** [comment]
hi @cortisiko I am getting the following errors and cannot build the app with `yarn setup && yarn start:android:qa` is it working on your end?


image

image

**Comment by cortisiko** [comment]
hey hey @seaona, I see that you are building the QA version of the app to run tests. Detox does not run test against the QA version of the app but instead it runs tests against the regular debug version of the app. You can kick off the tests by running: `detox build -c android.emu.debug && detox test -c android.emu.debug --testNamePattern="Smoke"`

**Comment by seaona** [comment]
thank you @cortisiko ! 
As side note, the QA build script is fixed if this is installed previously in Linux `sudo apt-get install libicu-dev`

**Comment by cortisiko** [comment]
UPDATE I used my MacBook running the M1 chip 

All of the tests in the smoke tag are passing on android 
image


All of the tests in the regression tag are passing on android 
image

File: android/build.gradle [hunk 1]
```diff
@@ -24,5 +24,11 @@ buildscript {
 		classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlinVersion"
     	classpath("com.facebook.react:react-native-gradle-plugin")
 	}
+	allprojects { // (4)
```
Comment (leotm) [inline android/build.gradle:27]: can we remove this comment
or expand upon the numéro 4

**Comment by leotm** [comment]
just posting initial local `detox build -c android.emu.debug` time for reviewers to get an idea of how long to wait 🍵☕ 

Screenshot 2023-06-29 at 13 03 39

File: e2e/pages/modals/NetworkEducationModal.js [hunk 2]
```diff
@@ -7,9 +7,19 @@ import {
...
 export default class NetworkEducationModal {
   static async tapGotItButton() {
-    await TestHelpers.waitAndTap(NETWORK_EDUCATION_MODAL_CLOSE_BUTTON_ID);
+    if (device.getPlatform() === 'ios') {
+      await TestHelpers.tap(NETWORK_EDUCATION_MODAL_CLOSE_BUTTON_ID);
+    } else {
+      await TestHelpers.tapByText('Got it');
+
+      // await TestHelpers.tapByDescendentTestID(
+      //   NETWORK_EDUCATION_MODAL_CONTAINER_ID,
```
Comment (Andepande) [inline e2e/pages/modals/NetworkEducationModal.js:19]: Should we remove commented items or are they TODO's?

**Review by Andepande — DISMISSED** [review]
LG

File: android/build.gradle (hunk 1 above)
Comment (cortisiko) [reply android/build.gradle]: Yeah, the 4 signified nothing really. It was just a placeholder comment I left for myself. Updated!

File: e2e/pages/modals/NetworkEducationModal.js (hunk 2 above)
Comment (cortisiko) [reply e2e/pages/modals/NetworkEducationModal.js]: Good eye. Definitely not TODOs. I have removed the commented code.

File: android/build.gradle [hunk 3]
```diff
@@ -8,7 +8,7 @@ buildscript {
 		compileSdkVersion = 33
 		targetSdkVersion = 33
 		// We use NDK 23 which has both M1 support and is the side-by-side NDK version from AGP.
-		ndkVersion = "24.0.8215888"
```
Comment (leotm) [inline android/build.gradle:11]: @jpcloureiro reflecting back, do we remember why we bumped ahead of official/current `23.1.7779620` for the [upgrade](https://github.com/MetaMask/metamask-mobile/pull/6220/files#r1247879219)?

File: android/build.gradle [hunk 4]
```diff
@@ -8,7 +8,7 @@ buildscript {
...
+		ndkVersion = "23.1.7779620"
```
Comment (leotm) [inline android/build.gradle:11]: did we downgrade from `24.0.8215888` due to failing tests?

Comment (leotm) [reply android/build.gradle:11]: this change to v23 looks correct after checking
- https://react-native-community.github.io/upgrade-helper/?from=0.66.5&to=0.71.6#RnDiffApp-android-build.gradle
- https://github.com/facebook/react-native/blob/v0.71.6/template/android/build.gradle#L11

File: android/build.gradle (hunk 3 above)
Comment (leotm) [reply android/build.gradle:11]: since 23 looks [correct](https://github.com/MetaMask/metamask-mobile/pull/6384/files#r1247889384)

Comment (jpcloureiro) [reply android/build.gradle:11]: I believe this change came from [another PR](https://github.com/MetaMask/metamask-mobile/pull/6327/commits/2ddcff66a121283cc2633b34dc0ea224899dc9f0) to address bitrise builds on M1 machines.

@sethkfman is it safe to bring back ndk to ver 23? Does bitrise support it?

Comment (leotm) [reply android/build.gradle:11]: and spotted [24.0.8215888](https://github.com/android/ndk/wiki/Unsupported-Downloads#r24) been the latest unsupported version awhile

Comment (leotm) [reply android/build.gradle:11]: thx @jpcloureiro ^ can see that was from couple months ago

@cortisiko i presume we're good to proceed on `23.1.7779620` after officially passing QA 🔥

File: android/build.gradle (hunk 4 above)
Comment (leotm) [reply android/build.gradle:11]: resolving convo ^ thx @jpcloureiro for answering in other [thread](https://github.com/MetaMask/metamask-mobile/pull/6384/files#r1247893326)

and confirmed v23 is the officially supported latest correct NDK version

with current QA is passing

File: android/build.gradle (hunk 3 above)
Comment (leotm) [reply android/build.gradle:11]: can confirm `24.0.8215888` removed from official supported list
https://developer.android.com/studio/projects/install-ndk

and `23.1.7779620` is the side-by-side NDK version from our correct [AGP 7.3.x](https://github.com/MetaMask/metamask-mobile/pull/6384/files#diff-197b190e4a3512994d2cebed8aff5479ff88e136b8cc7a4b148ec9c3945bd65aR23)

this ^ guarantees better build stability and runtime stability/perf

**Comment by leotm** [comment]
in `bitrise.yml` we're still setting [`NDK_VERSION: 24.0.8215888`](https://github.com/MetaMask/metamask-mobile/pull/6384/files#diff-61d1ee831cb1df242161a7d23171914a1d5d628c7659b52db957d80a9394bd68L702)

```yml
app:
  envs:
    - ...
    - opts:
        is_expand: false
      NDK_VERSION: 24.0.8215888
    - ...
```

for (step)

```yml
  ...
  android_e2e_test:
    ...
    steps:
      - restore-gradle-cache@1: {}
      - install-missing-android-tools@3:
          inputs:
            - ndk_version: $NDK_VERSION
            - gradlew_path: $PROJECT_LOCATION/gradlew
          is_always_run: false
      - ...
```

so  [install-missing-android-tools](https://github.com/bitrise-steplib/steps-install-missing-android-tools) for `24.0.8215888`
despite our app now on `23.1.7779620` with QA success :suspect: 

is this step still required? if not let's remove it
if yes let's update our bitrise env var to`NDK_VERSION: 23.1.7779620` to install correct missing tools

File: android/build.gradle (hunk 3 above)
Comment (leotm) [reply android/build.gradle:11]: ah you beat me to it (; https://github.com/MetaMask/metamask-mobile/pull/6384#issuecomment-1614836043

**Review by leotm — COMMENTED** [review]
closing thought (after prev comments)

from our BitRise [XL](https://github.com/MetaMask/metamask-mobile/assets/1881059/a5eec2e6-d96d-4927-9f11-32cafe20e8c4) machine stack

```log
Android & Docker, on Ubuntu 20.04
https://github.com/bitrise-io/bitrise.io/blob/master/system_reports/LINUX/linux-docker-android-20.04.log

...
$ cat /opt/android-sdk-linux/ndk-bundle/source.properties
Pkg.Desc = Android NDK
Pkg.Revision = 23.1.7779620

$ tree -L 2 /opt/android-sdk-linux/ndk-bundle
/opt/android-sdk-linux/ndk-bundle
├── build
│   ├── cmake
│   ├── ...
...
```

we're configured on NDK `23.1.7779620` by default
it appears with correct `cmake` binary from the provided `ndk-bundle/build`

so wondering why manually grabbing `cmake` _3.22.1_ via sdk manager is needed? (https://github.com/MetaMask/metamask-mobile/commit/a69fe58f5cb3224c78b40a2259d798e5ffbf9156)
_nb: this [diff](https://github.com/MetaMask/metamask-mobile/pull/6220/files#diff-61d1ee831cb1df242161a7d23171914a1d5d628c7659b52db957d80a9394bd68R300-R306) is reflected in the upgrade PR_

if the additional step is deffo needed
i initially expected it would be fetched/unpacked/[etc](https://github.com/leotm/react-native-template-new-architecture/pull/915/files#diff-b46b6cdc887427355980140454172d7b5ff3dd1f68523667bdd4947358d927d8L59-L68) via our `install-missing-android-tools` step
but after further [digging](https://github.com/bitrise-steplib/steps-install-missing-android-tools/issues/41#issuecomment-462839734) can see we're right to do this (most simply) via sdk manager 👍

**Review by Cal-L — DISMISSED** [review]
LGTM

**Review by Cal-L — APPROVED** [review]
LGTM

**Pull request merged**